sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala

```diff
@@ -205,6 +205,10 @@ object AnsiTypeCoercion extends TypeCoercionBase {
       case (StringType, AnyTimestampType) =>
         Some(AnyTimestampType.defaultConcreteType)
 
+      // If the StringType has any collation other than UTF8_BINARY, it shouldn't be implicitly
+      // cast to StringType with collation 0.
```

Comment:
looks the code does not follow what the comment says?

Reply:
perhaps the explanation is unclear, see how this sounds